### PR TITLE
other functions from musl

### PR DESCRIPTION
**Pull request by burrbull**
Absent functions added by @m1el in https://github.com/m1el/libm/tree/lots-of-untested-ports

### CODE REVIEW

**Comment by m1el** [comment]
Uh, these weren't tested and are probably broken.
I should organize this as separate pull requests, because all of these were commited ad-hoc.

**Comment by burrbull** [comment]
Ok. I'll close this PR in favour of yours.
You can run tests with `cargo test --features="musl-reference-tests"`.
Most of functions works fine.

File: build.rs [hunk 1]
```diff
@@ -41,6 +41,19 @@ mod musl_reference_tests {
         "rem_pio2.rs",
         "rem_pio2_large.rs",
         "rem_pio2f.rs",
+        "remquo.rs",
+        "remquof.rs",
+        "lgamma.rs",  // lgamma passed, lgamma_r has more than 1 result
```
Comment (alexcrichton) [inline build.rs]: Could support for testing these functions that have more than one result be added? I think it'd be better than ignoring all of the files here

File: build.rs [hunk 2]
```diff
@@ -41,6 +41,19 @@ mod musl_reference_tests {
...
+        "lgammaf.rs", // lgammaf passed, lgammaf_r has more than 1 result
+        "frexp.rs",   // more than 1 result
+        "frexpf.rs",  // more than 1 result
+        "sincos.rs",  // more than 1 result
+        "sincosf.rs", // more than 1 result
+        "modf.rs",    // more than 1 result
+        "modff.rs",   // more than 1 result
+        "asinef.rs",  // not exists
```
Comment (alexcrichton) [inline build.rs]: What does "not exists" mean here?

File: build.rs [hunk 3]
```diff
@@ -41,6 +41,19 @@ mod musl_reference_tests {
         "rem_pio2.rs",
         "rem_pio2_large.rs",
         "rem_pio2f.rs",
+        "remquo.rs",
+        "remquof.rs",
```
Comment (alexcrichton) [inline build.rs]: Something I should have done last time is add a comment here, but could a comment be added to these files indicating that they don't have public APIs and are only considered to be implementation details?

File: build.rs [hunk 4]
```diff
@@ -41,6 +41,19 @@ mod musl_reference_tests {
...
+        "lgammaf.rs", // lgammaf passed, lgammaf_r has more than 1 result
+        "frexp.rs",   // more than 1 result
+        "frexpf.rs",  // more than 1 result
+        "sincos.rs",  // more than 1 result
+        "sincosf.rs", // more than 1 result
+        "modf.rs",    // more than 1 result
+        "modff.rs",   // more than 1 result
+        "asinef.rs",  // not exists
+        "jn.rs",      // passed, but very slow
+        "jnf.rs",     // passed, but very slow
```
Comment (alexcrichton) [inline build.rs]: Can you comment on the slowness? How so? Could the number of tests simply be reduced?

(I would prefer to ignore literally nothing that is a public interface of this library)

File: build.rs (hunk 1 above)
Comment (burrbull) [reply build.rs]: Of course, improving test system is better. But I'm not sure I can do it.

File: build.rs (hunk 3 above)
Comment (burrbull) [reply build.rs]: Ok.

File: build.rs (hunk 4 above)
Comment (burrbull) [reply build.rs]: About 10 minutes with NTESTS = 5

```
test ynf_matches_musl ... ok
test jn_matches_musl ... test jn_matches_musl has been running for over 60 seconds
test jnf_matches_musl ... test jnf_matches_musl has been running for over 60 seconds
test jn_matches_musl ... ok
test jnf_matches_musl ... ok
```

File: build.rs (hunk 2 above)
Comment (burrbull) [reply build.rs]: Panics with
```
--- stderr
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-Wl,--eh-frame-hdr" "-m64" "-nostdlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/crt1.o" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/crti.o" "-L" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "gen.gen.7rcbfp3g-cgu.0.rcgu.o" "gen.gen.7rcbfp3g-cgu.1.rcgu.o" "gen.gen.7rcbfp3g-cgu.10.rcgu.o" "gen.gen.7rcbfp3g-cgu.11.rcgu.o" "gen.gen.7rcbfp3g-cgu.12.rcgu.o" "gen.gen.7rcbfp3g-cgu.13.rcgu.o" "gen.gen.7rcbfp3g-cgu.14.rcgu.o" "gen.gen.7rcbfp3g-cgu.15.rcgu.o" "gen.gen.7rcbfp3g-cgu.2.rcgu.o" "gen.gen.7rcbfp3g-cgu.3.rcgu.o" "gen.gen.7rcbfp3g-cgu.4.rcgu.o" "gen.gen.7rcbfp3g-cgu.5.rcgu.o" "gen.gen.7rcbfp3g-cgu.6.rcgu.o" "gen.gen.7rcbfp3g-cgu.7.rcgu.o" "gen.gen.7rcbfp3g-cgu.8.rcgu.o" "gen.gen.7rcbfp3g-cgu.9.rcgu.o" "-o" "gen" "gen.ml8rp173tno6czu.rcgu.o" "-Wl,--gc-sections" "-no-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-e9655bcee52420f6.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libpanic_unwind-94a79f5396bbf06e.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libbacktrace_sys-7a0a2cd8401e3fbb.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_demangle-544fe11ac5351948.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libhashbrown-2de4e4f6ac9ab0a9.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_alloc-9b193946d0ee39d9.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libunwind-c94045c5315cd4de.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liblibc-1e84ab9d7a609c9b.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liballoc-d698a486b5d0cf97.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_core-bf2642637f4f98fd.rlib" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcore-82c12243ccd26b23.rlib" "-Wl,--end-group" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcompiler_builtins-0bd15cf37b00baa7.rlib" "-static" "-Wl,-Bdynamic" "/home/burrbull/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/crtn.o"
  = note: gen.gen.7rcbfp3g-cgu.0.rcgu.o: In function `gen::main':
          gen.7rcbfp3g-cgu.0:(.text._ZN3gen4main17heffb9d4db471d1caE+0x58aa): undefined reference to `asinef'
          collect2: error: ld returned 1 exit status

```

File: build.rs (hunk 1 above)
Comment (burrbull) [reply build.rs]: Stop. I forgot that these function have other signature in C:
`void sincos(double x, double *sin, double *cos)`
whereas in Rust we use
`fn sincos(x: f64) -> (f64, f64)`

File: build.rs (hunk 2 above)
Comment (alexcrichton) [reply build.rs]: If the function doesn't exist in musl why are we exporting it from this library?

File: build.rs (hunk 1 above)
Comment (alexcrichton) [reply build.rs]: I am personally wary enough of these implementations that I do not want to land this PR without full tests for all added functions.

File: build.rs (hunk 4 above)
Comment (alexcrichton) [reply build.rs]: With only 5 tests it takes 5 minutes to run these tests? Does that mean these are just too slow to land?

File: build.rs (hunk 1 above)
Comment (burrbull) [reply build.rs]: I can add individual random tests for these functions to the files where they are implemented.

File: build.rs (hunk 4 above)
Comment (burrbull) [reply build.rs]: `jn(n: i32, mut x: f64) -> f64` is Bessel function where `n` is order.
I `n` in think tests for `jn` and `jnk` should be limited to small values like 0, 1, .. 5.

File: build.rs (hunk 2 above)
Comment (burrbull) [reply build.rs]: I'll remove this one.

File: build.rs (hunk 3 above)
Comment (m1el) [reply build.rs]: These are supposed to be public APIs.

**Comment by burrbull** [comment]
Test system has been modified and works with all public functions except sincos.

**Comment by alexcrichton** [comment]
I pushed up a commit to remove a number of exceptions and additonally add support for testing sincos. I also added better generation of random floats and new ilogb/ilogbf functions are now failing tests. I randomly commented out a small block in sincos/sincosf and it continues to pass the randomized tests which sort of worries me as well...

**Comment by alexcrichton** [comment]
I've also pushed up support for jn/jnf. As I've mentioned before I don't want to land anything that's untested. This is a critical library which is basically always assumed to be correct, so CI is close to the last line of defense for correctness.

File: build.rs [hunk 5]
```diff
@@ -143,15 +141,28 @@ mod musl_reference_tests {
...
+            let mut inputs = function
+                .args
+                .iter()
+                .map(|ty| ty.gen_i64(rng))
+                .collect::<Vec<_>>();
+
+            // First argument to this function appears to be a number of
+            // iterations, so passing in massive random numbers causes it to
+            // take forever to execute, so make sure we're not running random
+            // math code until the heat death of the universe.
```
Comment (burrbull) [inline build.rs]: :smile:

**Comment by alexcrichton** [comment]
👍

**Pull request merged**